dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/JdbcEventCommentStore.java

```diff
@@ -0,0 +1,199 @@
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import static java.util.stream.Collectors.toList;
+
+import java.sql.PreparedStatement;
+import java.util.List;
+
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.trackedentitycomment.TrackedEntityComment;
+import org.springframework.dao.DataAccessException;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * @author Giuseppe Nespolino <g.nespolino@gmail.com>
+ */
+@Repository
+@RequiredArgsConstructor
+@Slf4j
+public class JdbcEventCommentStore
+{
+
+    private final JdbcTemplate jdbcTemplate;
+
+    private final String INSERT_EVENT_NOTE_SQL = "INSERT INTO TRACKEDENTITYCOMMENT (trackedentitycommentid, " + // 0
+        "uid, " + // 1
+        "commenttext, " + // 2
+        "created, " + // 3
+        "creator," + // 4
+        "lastUpdated" + // 5
+        ")  values ( nextval('hibernate_sequence'), ?, ?, ?, ?, ?)";
+
+    private final static String INSERT_EVENT_COMMENT_LINK = "INSERT INTO programstageinstancecomments (programstageinstanceid, "
+        + "sort_order, trackedentitycommentid) values (?, ?, ?)";
+
+    /**
+     * Save all the comments ({@see TrackedEntityComment} for the list of
+     * {@see ProgramStageInstance}
+     *
+     * @param batch a List of {@see ProgramStageInstance}
+     */
+    void saveAllComments( List<ProgramStageInstance> batch )
+    {
+        try
+        {
+            // List of PSI that has at least one non empty comment (i.e. PSO having comments
+            // that can actually be saved)
+            // In resulting PSI list, all comments without text are removed.
+            List<ProgramStageInstance> programStageInstances = batch.stream()
+                .map( this::withoutEmptyComments )
```

Comment:
Is `.map` necessary here? Wouldn't `batch.stream().filter( this:hasComments )...` achieve the same?

Reply:
this is to ensure that program instances only have comments with some text, so we don't have to check it again later when storing.